# Matemáticas para Inteligencia Artificial (I)




### 1. Normas, distancias y ángulos




Sea $V$ un $\mathbb{R}$-espacio vectorial donde se ha definido un producto escalar $\langle\cdot,\cdot\rangle$. Recordemos que para que $\langle \cdot , \cdot\rangle$ sea un producto escalar, debe cumplir las siguientes propiedades: 

- $\langle x,y\rangle=\langle y,x\rangle, \; \forall\, x,y\in V$
- $\langle \alpha x + \beta y,z\rangle=\alpha \langle x,z\rangle+\beta\langle y, z\rangle, \; \forall\, x,y,z\in V$, $\forall\, \alpha,\beta\in\mathbb{R}$
- $\langle x,x\rangle \geq 0$, $\forall\, x\in V$.
- $\langle x,x\rangle =0$ si y sólo si $x=0$. 


Si se ha definido un producto escalar en $V$, entonces podemos calcular:

- normas de elementos $v\in V$ como $||v||:=\sqrt{\langle v,v\rangle}$
- distancia entre dos elementos $u,v\in V$ como $d(u,v):=||u-v||$
- ángulo $\alpha$ que forman dos elementos $u,v \in V$ a través de $\operatorname{cos}(\alpha):=\dfrac{|\langle u,v\rangle|}{||u||\cdot||v||}\in[-1,1]$

Veamos dos maneras alternativas de calcular el producto escalar de dos vectores en $V=\mathbb{R}^3$.

In [1]:
import numpy as np
from numpy import linalg as la

u=np.array([[1,0,1]])
v=np.array([[1],[2],[3]])

print("u =", u, "\n")
print("v =", v)

u = [[1 0 1]] 

v = [[1]
 [2]
 [3]]


In [3]:
print(np.sum(u*v.T), "\n")
print(u@v)

4 

[[4]]


In [5]:
print("La distancia entre u y v es", la.norm(u-v.T))
print("El ángulo que forman u y v es", (np.abs(u@v)/(la.norm(u)*la.norm(v)))[0][0])

La distancia entre u y v es 2.8284271247461903
El ángulo que forman u y v es 0.7559289460184544


Además de la distancia euclídea, existen otros métodos para investigar cuándo dos vectores de $\mathbb{R}^n$ son _cercanos_ o _similares_. Por ejemplo, cuando la longitud de los vectores no es importante, la **similitud del coseno** resulta muy útil. Esto ocurre por ejemplo en _text mining_. 


$$ \text{sim}_{(u, v)}=\dfrac{\langle u, v\rangle}{||u|| ·||v||}=\text{cos}(\text{ángulo}(u, v))\in[-1,1],$$ 


<img src=https://lh6.googleusercontent.com/ycXLRE6YUiFrVJpkYXJe8I7oCiuUSYLhfBHvn81N3_AARdiEuswYKLqC5mLNqdqTiAkfCN7hBBdrIgQi6OAbOshJE1d3q_0XuWjik_KaQqryrh63PJiS9wDwT0M0NZ_AbOFpvEZgcMjbgYBXtM3VVdM width="1000">


A modo de ejemplo, veamos la _similitud del coseno_ entre los dos vectores anteriores.

In [7]:
print(np.abs((u@v)[0][0])/(la.norm(u)*la.norm(v)))

0.7559289460184544


Veamos ahora cuán similares son $v$ y el vector $(1,2,4)$.

In [9]:
w=np.array([[1,2,4]])
print(np.abs((w@v)[0][0])/(la.norm(v)*la.norm(w)))

0.9914601339836675


En este contexto, también se pueden definir otras distancias, como por ejemplo:


<img src=https://miro.medium.com/v2/resize:fit:1400/1*vAtQZbROuTdp36aQQ8cqBA.png width="800">


La célebre **distancia de Manhattan**, que se define como $d(u,v):=\displaystyle\sum_i |u_i-v_i|$, aplicada a $u$ y $v$, y a $v$ y $w$ proporcionaría:

In [11]:
print(np.sum(np.abs(u-v.T)))
print(np.sum(np.abs(w-v.T)))

4
1


Por tanto, lo _cercanos_ que sean dos vectores/objetos depende de la distancia que se considere.

### 2. Algoritmos de recomendación


Un **sistema de recomendación** es un algoritmo que nos permite dar predicciones de cuál es el producto o ítem más adecuado para un usuario. Los sistemas de recomendación pueden ser de varias clases según el algoritmo utilizado, aunque solo nos centraremos en la presente práctica en los de _filtrado colaborativo_. 


Los sistemas de recomendación basados en algoritmos de **filtrado colaborativo** utilizan las valoraciones o interacciones de los usuarios sobre ciertos elementos del conjunto de productos, con el objetivo de predecir valoraciones o interacciones en el resto de los elementos y recomendar los de mayor valoración predicha. Ejemplos en esta línea pueden ser **Netflix**, **Spotify**, **Youtube**, etc.


Los datos iniciales de los que disponemos son una base de datos de usuarios o clientes $(user_1, user_2, \ldots, user_M)$ a los que vamos a recomendar una serie de productos o ítems $(ítem_1, ítem_2,\ldots, ítem_N)$, y las interacciones o puntuaciones de esos usuarios sobre algunos de los ítems, matriz $L$ de dimensión $M\times N$. 


**Ejemplo.** Supongamos que 10 usuarios están buscando en la página de **Ryanair** en el último mes vuelos desde Madrid (MAD) a Málaga (AGP), Las Palmas de Gran Canaria (LPA), Sevilla (SVQ), Valencia (VLC), Ibiza (IBZ) y Santiago de Compostela (SCQ). Supongamos que la fila $i$ del fichero `vuelos.csv` representa el número de búsquedas que ha realizado el usuario $i$ sobre cada una de las rutas antes mencionadas (en el orden citado). 

In [13]:
import pandas as pd

vuelos = pd.read_csv('./vuelos.csv')

In [15]:
vuelos.head()

MAD-AGP  MAD-LPA  MAD-SVQ  MAD-VLC  MAD-IBZ  MAD-SCQ
0      0.0      0.0      1.0      0.0      0.0      2.0
1      0.0      2.0      1.0      1.0      0.0      0.0
2      0.0      0.0      1.0      1.0      1.0      1.0
3      0.0      0.0      1.0      2.0      0.0      0.0
4      0.0      0.0      1.0      1.0      0.0      0.0

In [17]:
V=vuelos.to_numpy()
print(V)

[[0. 0. 1. 0. 0. 2.]
 [0. 2. 1. 1. 0. 0.]
 [0. 0. 1. 1. 1. 1.]
 [0. 0. 1. 2. 0. 0.]
 [0. 0. 1. 1. 0. 0.]
 [0. 0. 2. 0. 0. 1.]
 [1. 0. 0. 1. 0. 1.]
 [0. 1. 0. 1. 0. 2.]
 [1. 2. 0. 0. 0. 1.]
 [1. 0. 1. 0. 1. 0.]]


**Ejercicio.** A vista de los datos anteriores:
- ¿qué vuelo le recomendarías a los usuarios 1 y 2? 
- ¿Y a los usuarios 3, 5, 7 y 10? 


Dos filtrados colaborativos célebres son:


1. User-user: personas con intereses similares en el pasado es probable que tengan intereses similares en el futuro. Por tanto, a la hora de recomendar a un usuario un ítem, nos fijamos en los intereses en los que se han interesado usuarios similares a él .


**Ejemplo.** Messi ha visto Juego de Tronos y Breaking Bad. Cristiano ha visto Juego de Tronos, Breaking Bad y Vikingos. El algoritmo detecta que Messi y Cristiano son usuarios similares, luego el sistema recomienda Vikingos a Messi.


2. Ítem-ítem: si a un usuario le ha interesado en el pasado un producto, es probable que en el futuro le interesen productos similares. Por tanto, a la hora de recomendar a un usuario un ítem, nos fijamos en ítems similares a los que se ha interesado en el pasado.


**Ejemplo.** Messi, Cristiano, Haaland y Mbappé han visto Origen. Messi, Cristiano y Haaland también han visto Shutter Island. El algoritmo detecta que Origen y Shutter Island son productos similares, luego el sistema recomienda a Mbappé ver Shutter Island.


Queda de manifiesto, por tanto, que la clave es conocer la similitud entre usuarios o ítems, siendo la opción más "manejable" aquella que tenga menos elementos. Observemos que no estamos interesados en la longitud de los vectores, sino simplemente en su similitud, luego resulta natural utilizar la similitud del coseno.



En nuestro ejemplo vamos a realizar un algoritmo de recomendación ítem-ítem con el objetivo de recomendar a cada usuario un vuelo y después responderemos de nuevo a las preguntas anteriores. Comencemos analizando la similitud del resto de vuelos con el vuelo 6, pues es el que más veces ha buscado el usuario 1.

In [19]:
print("sim(vuelo_1, vuelo_6) = ", (V[:,0:1].T@V[:,5:6])[0][0]/(la.norm(V[:,0:1])*la.norm(V[:,5:6])))

sim(vuelo_1, vuelo_6) =  0.33333333333333337


In [21]:
print("sim(vuelo_5, vuelo_6) = ", (V[:,4:5].T@V[:,5:6])[0][0]/(la.norm(V[:,4:5])*la.norm(V[:,5:6])))

sim(vuelo_5, vuelo_6) =  0.20412414523193148


Ahora hagámoslo en general, es decir, calculemos las similitudes entre todos los vuelos entre sí. Esto nos derá una matriz $S$ de similitud entre vuelos, que ha de ser simétrica y con unos en la diagonal.

In [23]:
S=(V/la.norm(V, axis=0)).T@(V/la.norm(V, axis=0))
print(S)

[[1.         0.38490018 0.18257419 0.19245009 0.40824829 0.33333333]
 [0.38490018 1.         0.21081851 0.33333333 0.         0.38490018]
 [0.18257419 0.21081851 1.         0.52704628 0.4472136  0.45643546]
 [0.19245009 0.33333333 0.52704628 1.         0.23570226 0.38490018]
 [0.40824829 0.         0.4472136  0.23570226 1.         0.20412415]
 [0.33333333 0.38490018 0.45643546 0.38490018 0.20412415 1.        ]]


Sabiendo que el usuario 1 ha buscado previamente los vuelos 3 y 6 (una y dos veces, respectivamente), necesitamos buscar un vuelo que sea lo más parecido posible a los vuelos 3 y 6 (y preferiblemente más al 6 que al 3, es decir, **debemos ponderar estas búsquedas**). Por tanto, si multiplicamos (producto escalar) cada fila de $S$ por la primera fila de $V$ obtendremos unos valores que representan la _recomendación_ de los vuelos. Por tanto, bastará con quedarse con el valor más alto.

In [25]:
V[0:1,:]@S

array([[0.84924085, 0.98061887, 1.91287093, 1.29684664, 0.85546189,
        2.45643546]])

Ahora solamente tenemos que ver el valor más alto de los anteriores y que corresponda con un cero de la primera fila de $V$ (i.e. con un vuelo que aún no haya buscado). En este caso, sería el vuelo 4. Responde a las preguntas anteriores nuevamente fijándote en la siguiente matriz.

In [27]:
print(np.round(V@S, 3))
print(V)

[[0.849 0.981 1.913 1.297 0.855 2.456]
 [1.145 2.544 1.949 2.194 0.683 1.611]
 [1.117 0.929 2.431 2.148 1.887 2.045]
 [0.567 0.877 2.054 2.527 0.919 1.226]
 [0.375 0.544 1.527 1.527 0.683 0.841]
 [0.698 0.807 2.456 1.439 1.099 1.913]
 [1.526 1.103 1.166 1.577 0.848 1.718]
 [1.244 2.103 1.651 2.103 0.644 2.77 ]
 [2.103 2.77  1.061 1.244 0.612 2.103]
 [1.591 0.596 1.63  0.955 1.855 0.994]]
[[0. 0. 1. 0. 0. 2.]
 [0. 2. 1. 1. 0. 0.]
 [0. 0. 1. 1. 1. 1.]
 [0. 0. 1. 2. 0. 0.]
 [0. 0. 1. 1. 0. 0.]
 [0. 0. 2. 0. 0. 1.]
 [1. 0. 0. 1. 0. 1.]
 [0. 1. 0. 1. 0. 2.]
 [1. 2. 0. 0. 0. 1.]
 [1. 0. 1. 0. 1. 0.]]


Trabajemos ahora con otro ejemplo de mayores dimensiones. Concretamente, la siguiente base de datos contiene puntuaciones de usuarios sobre ciertas películas.

In [29]:
import pandas as pd

pelis = pd.read_csv('./ratings_example.csv')
pelis.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

Lo primero que debemos hacer en este caso es construirnos la matriz que nos interesa, donde las filas sean los usuarios y las columnas sean las películas.

In [31]:
pelis_matriz=pelis.pivot_table(index='userId', columns='movieId', values='rating').fillna(0)
pelis_matriz.head()

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
3           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
5           0.0     0.0     4.0     0.0     0.0     0.0     0.0     0.0   

movieId  9       10      ...  161084  161155  161594  161830  161918  161944  \
userId                   ...                                                   
1           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
2           0.0     4.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
3           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
4           0.0     4.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
5           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

movieId  162376  162542  162672  163949  
userId                                   
1           0.0     0.0     0.0     0.0  
2           0.0     0.0     0.0     0.0  
3           0.0     0.0     0.0     0.0  
4           0.0     0.0     0.0     0.0  
5           0.0     0.0     0.0     0.0  

[5 rows x 9066 columns]

In [33]:
M=pelis_matriz.to_numpy()
print(M ,"\n", M.shape)

[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [4. 0. 0. ... 0. 0. 0.]
 [5. 0. 0. ... 0. 0. 0.]] 
 (671, 9066)


Por tanto, tenemos 671 usuarios que han valorado 9066 películas (observemos que hay etiquetas correspondientes a películas que no han sido vistas por nadie, como por ejemplo las que hay entre la 162673 y la 163948, luego no aparecen como columnas). 


En este caso, tenemos menos usuarios que películas, luego nos interesa ver la similitud entre usuarios.

In [35]:
print("sim(user_1, user_2) = ", (M[0:1,:]@M[1:2,:].T)[0][0]/(la.norm(M[0:1,:])*la.norm(M[1:2,:])))

sim(user_1, user_2) =  0.0


In [37]:
print("sim(user_1, user_4) = ", (M[0:1,:]@M[3:4,:].T)[0][0]/(la.norm(M[0:1,:])*la.norm(M[3:4,:])))

sim(user_1, user_4) =  0.07448244558021601


Ahora hagámoslo en general, es decir, calculemos las similitudes entre todos los usuarios.

In [39]:
S=(M.T/la.norm(M.T, axis=0)).T@(M.T/la.norm(M.T, axis=0))
print(S)

[[1.         0.         0.         ... 0.06291708 0.         0.01746565]
 [0.         1.         0.12429498 ... 0.02413984 0.17059464 0.1131753 ]
 [0.         0.12429498 1.         ... 0.08098382 0.13660585 0.17019275]
 ...
 [0.06291708 0.02413984 0.08098382 ... 1.         0.04260878 0.08520194]
 [0.         0.17059464 0.13660585 ... 0.04260878 1.         0.22867673]
 [0.01746565 0.1131753  0.17019275 ... 0.08520194 0.22867673 1.        ]]


Si multiplicamos ahora el primer vector fila de $S$ por cada una columna de $M$, obtendremos un número que indica la _recomendación_ de esa película para el usuario 1. Y, en general, si multiplicamos por todas las columnas de $M$ y nos quedamos con el valor más alto (de aquellas películas que no ha visto el usuario 1), esa será nuestra recomendación.

In [41]:
S[0:1,:]@M

array([[20.87571921,  8.46503375,  4.44940719, ...,  0.10949069,
         0.06569442,  0.28836607]])

In [43]:
r=np.argmax(S[0:1,:]@M)
print(r)
print((S[0:1,:]@M)[0][r])
print((S[0:1,:]@M)[0][r]==np.max(S[0:1,:]@M))

232
32.148441326631
True


In [45]:
print(M[0,r])

0.0


In [47]:
recomen=(S[0:1,:]@M)[0]
print(recomen)

recomen[M[0]!=0]=0
print(recomen)

# para asegurarnos de que coge el valor máximo de aquellas películas que el usuario no ha visto

r=np.argmax(recomen)
print(r)    

# seleccionamos ahora el mayor valor, y entonces seguro que no la ha visto

[20.87571921  8.46503375  4.44940719 ...  0.10949069  0.06569442
  0.28836607]
[20.87571921  8.46503375  4.44940719 ...  0.10949069  0.06569442
  0.28836607]
232


Luego le recomendamos la película/columna 232 de $M$, que es la película con Id 260:

In [49]:
peliculas=pelis.movieId.to_numpy()

print("peliculas =", peliculas)

peliculas = [  31 1029 1061 ... 6365 6385 6565]


In [51]:
pelis_matriz.columns[r]

260

Si queremos ahora saber qué recomendación haríamos a cada usuario, basta con hacer lo anterior en general: multiplicamos $S$ por $M$ y ordenaríamos los elementos de la misma forma anterior.

In [53]:
K=S@M

In [55]:
def Recomendacion(i):
    K[i-1][M[i-1]!=0]=0
    return np.argmax(K[i-1])

In [57]:
for i in range(671):
    print("Al usuario", i+1, "le recomendamos la película", np.unique(peliculas)[Recomendacion(i+1)], "\n")

Al usuario 1 le recomendamos la película 260 

Al usuario 2 le recomendamos la película 318 

Al usuario 3 le recomendamos la película 260 

Al usuario 4 le recomendamos la película 318 

Al usuario 5 le recomendamos la película 318 

Al usuario 6 le recomendamos la película 318 

Al usuario 7 le recomendamos la película 296 

Al usuario 8 le recomendamos la película 480 

Al usuario 9 le recomendamos la película 296 

Al usuario 10 le recomendamos la película 260 

Al usuario 11 le recomendamos la película 318 

Al usuario 12 le recomendamos la película 296 

Al usuario 13 le recomendamos la película 593 

Al usuario 14 le recomendamos la película 260 

Al usuario 15 le recomendamos la película 595 

Al usuario 16 le recomendamos la película 296 

Al usuario 17 le recomendamos la película 1196 

Al usuario 18 le recomendamos la película 1 

Al usuario 19 le recomendamos la película 2571 

Al usuario 20 le recomendamos la película 589 

Al usuario 21 le recomendamos la película 318 

A